cats/cats-redis/src/main/java/com/netflix/spinnaker/cats/redis/cluster/ClusteredSortAgentScheduler.java

```diff
@@ -0,0 +1,297 @@
+        // for us.
+        AgentWorker worker = agents.get(key);
+        if (worker != null) {
+          releaseAgent(worker.agent);
+        }
+      }
+
+      // Now look for agents that have been in the queue for at least INTERVAL time.
+      List<String> keys = new ArrayList<>();
+      keys.addAll(jedis.zrangeByScore(WAITING_SET, "-inf", score(NOW)));
+      Set<AgentWorker> workers = new HashSet<>();
+
+      // Loop until we either run out of threads to use, or agents (which are keys) to run.
+      while (!keys.isEmpty() && runningAgents.map(Semaphore::tryAcquire).orElse(true)) {
+        String agent = keys.remove(0);
+
+        AgentWorker worker = agents.get(agent);
+        String score;
+        if (worker != null && (score = acquireAgent(worker.agent)) != null) {
+          // This score is used to determine if the worker thread running the agent is allowed to store its results.
+          // If on release of this agent, the scores don't match, this agent was rescheduled by a separate thread.
+          worker.setScore(score);
+          workers.add(worker);
+        }
+      }
+
+      for (AgentWorker worker : workers) {
+        Future s = agentWorkPool.submit(worker);
+        assert !s.isCancelled(); // This is only to appease findBugs
+      }
+    }
+  }
+
+  private static class AgentWorker implements Runnable {
+    private final Agent agent;
+    private final CachingAgent.CacheExecution agentExecution;
+    private final ExecutionInstrumentation executionInstrumentation;
+    private final ClusteredSortAgentScheduler scheduler;
+    private String acquireScore;
+
+    AgentWorker(Agent agent, CachingAgent.CacheExecution agentExecution, ExecutionInstrumentation executionInstrumentation, ClusteredSortAgentScheduler scheduler) {
+      this.agent = agent;
+      this.agentExecution = agentExecution;
+      this.executionInstrumentation = executionInstrumentation;
+      this.scheduler = scheduler;
+    }
+
+    public void setScore(String score) {
+      acquireScore = score;
+    }
+
+    @Override
+    public void run() {
+      assert acquireScore != null;
+      CacheResult result = null;
+      try {
+        executionInstrumentation.executionStarted(agent);
+        long startTime = System.nanoTime();
+        result = agentExecution.executeAgentWithoutStore(agent);
+        executionInstrumentation.executionCompleted(agent, TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startTime));
+      } catch (Throwable cause) {
+        executionInstrumentation.executionFailed(agent, cause);
+      } finally {
+        // Regardless of success or failure, we need to try and release this agent. If the release is successful (we
+        // own this agent), and a result was created, we can store it.
+        if (scheduler.conditionalReleaseAgent(agent, acquireScore) && result != null) {
+          agentExecution.storeAgentResult(agent, result);
+        }
+
+        scheduler.runningAgents.ifPresent(Semaphore::release);
```

Comment:
should do this before the call to conditionalReleaseAgent (which could fail if there is an error talking to redis) or alternatively do this inside conditionalReleaseAgent and keep all access to the semaphore inside the scheduler